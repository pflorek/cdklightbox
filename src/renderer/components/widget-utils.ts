import {
  AnyWidgetState,
  ConstructWidgetState,
  getRef,
  Parameter,
  StaticFunctionWidgetState,
  WidgetState,
  WorkbenchState,
} from '../state';
import { PickedWidget } from './Picker';

function getStackOrOpen(widgetState: WidgetState): WidgetState {
  let currentWidgetState = widgetState;
  const isStackClass = (forType: string) => forType === '@aws-cdk/core.Stack';

  while (!isStackClass(currentWidgetState.forType)) {
    currentWidgetState = currentWidgetState.scope as WidgetState;
  }
  return currentWidgetState;
}

function isStackOrOpen(widgetState: WidgetState) {
  return !!getStackOrOpen(widgetState);
}

export const handleValuePicked = (
  addWidget: (
    newWidget: WidgetState,
    refUpdater?: (state: WorkbenchState, newKey: string) => void
  ) => void,
  updateWidget: (
    updater: (
      widgets:
        | Record<string, WidgetState>
        | Record<string, StaticFunctionWidgetState>
        | Record<string, ConstructWidgetState>
    ) => void
  ) => void,
  removePicker: (picker: string) => void,
  widgetState: WidgetState,
  widgetKey: string,
  parametersProvider: (widgetState: AnyWidgetState) => Record<string, Parameter>
) => {
  return (props: PickedWidget) => {
    if (props.widget) {
      const newWidget = {
        ...props.widget,
        scope: isStackOrOpen(widgetState)
          ? widgetState.scope
          : getStackOrOpen(widgetState).scope,
      };
      addWidget(newWidget, (workbench: WorkbenchState, newKey: string) => {
        parametersProvider(workbench.widgets[widgetKey])[props.property.name] =
          {
            ref: getRef(newKey),
            widgetShapes: {},
          };
      });
    }
    if (props.value !== undefined) {
      updateWidget((widgets) => {
        const currentWidget = widgets[widgetKey];
        parametersProvider(currentWidget)[props.property.name] = {
          value: props.value,
          widgetShapes: {},
        };
      });
    }
    if (props.pickerKey) {
      removePicker(props.pickerKey);
    }
  };
};

export const handleValueZipped = (
  zip,
  unzip,
  constructKey: string,
  parametersProvider: (widgetState: AnyWidgetState) => Record<string, Parameter>
) => {
  return (prop: any, isZipped: boolean) => {
    if (isZipped) {
      zip(
        (state) => parametersProvider(state.widgets[constructKey])[prop.name]
      );
    } else {
      unzip(
        (state) => parametersProvider(state.widgets[constructKey])[prop.name]
      );
    }
  };
};
