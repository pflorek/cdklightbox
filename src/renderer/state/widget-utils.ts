import {
  EnumWidgetState,
  StaticFunctionWidgetState,
  WidgetState,
} from './states';
import { jsii } from '../jsii';

export function isStaticFunction(widgetState: StaticFunctionWidgetState) {
  return !!(widgetState.method && widgetState.parameters);
}

export function isEnum(widgetState: EnumWidgetState) {
  return !!(widgetState.method && widgetState.type === 'enum');
}

export function isConstruct(widgetState: WidgetState) {
  if ((widgetState as StaticFunctionWidgetState).type === 'staticFunction') {
    return false;
  }
  if ((widgetState as EnumWidgetState).type === 'enum') {
    return false;
  }
  const jsiiType = jsii.types[widgetState.forType];

  return !!(
    jsiiType.initializer?.parameters &&
    jsiiType.initializer?.parameters[0].type.fqn === 'constructs.Construct'
  );
}

export function isLooseObject(widgetState: WidgetState) {
  return (
    !isStaticFunction(widgetState as StaticFunctionWidgetState) &&
    !isEnum(widgetState as EnumWidgetState) &&
    !isConstruct(widgetState)
  );
}

export function getRef(key: string) {
  return `WIDGET#${key}`;
}

export enum WidgetTypeShape {
  LOOSE = 'LOOSE',
  STATIC_FUNCTION = 'STATIC_FUNCTION',
  ENUM = 'ENUM',
}

export function getKey(ref: string) {
  return ref.replace('WIDGET#', '');
}

export function isRef(key: string) {
  return key.startsWith('WIDGET#');
}
