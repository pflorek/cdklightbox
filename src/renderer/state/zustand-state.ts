import create from 'zustand';
import { persist, subscribeWithSelector } from 'zustand/middleware';
import { Set, WorkbenchState } from './states';
import { errorState } from './substates/error-state';
import { workingDirectoryState } from './substates/working-directory-state';
import { widgetsState } from './substates/widget-state';
import { statusState } from './substates/status-state';

// eslint-disable-next-line import/prefer-default-export
export const useWorkbenchStore = create<WorkbenchState>(
  // @ts-ignore
  subscribeWithSelector(
    persist(
      (set: Set) => ({
        ...errorState(set),
        ...statusState(set),
        ...workingDirectoryState(set),
        ...widgetsState(set),
        loadState(newState: Partial<WorkbenchState>) {
          set(() => newState);
        },
      }),

      {
        name: 'localStorage',
        getStorage: () => localStorage,
      }
    )
  )
);
