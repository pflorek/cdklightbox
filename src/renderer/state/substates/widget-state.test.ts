import {
  ConstructWidgetState,
  StaticFunctionWidgetState,
  WidgetState,
  WorkbenchState,
} from '../states';
import { getRef } from '../widget-utils';
import { useWorkbenchStore } from '../zustand-state';

const createStubWidgetState = () => ({
  id: '',
  x: 0,
  y: 0,
  forType: 'somefqnname',
});

describe('update position', () => {
  it('sets x and y', () => {
    const constructKey = 'whatever';
    useWorkbenchStore.setState({
      widgets: {
        [constructKey]: createStubWidgetState(),
      },
      pickers: {},
    });

    // when
    useWorkbenchStore.getState().updatePosition(constructKey, { x: 1, y: 2 });

    // then
    const updatedState: WorkbenchState = useWorkbenchStore.getState();
    expect(updatedState.widgets[constructKey].x).toEqual(1);
    expect(updatedState.widgets[constructKey].y).toEqual(2);
  });
});

describe('update construct id', () => {
  it('sets value', () => {
    const constructKey = 'whatever';
    useWorkbenchStore.setState({
      widgets: {
        [constructKey]: createStubWidgetState(),
      },
      pickers: {},
    });

    // when
    const expected = 'something';
    useWorkbenchStore.getState().updateConstructId(constructKey, expected);

    // then
    const updatedState: WorkbenchState = useWorkbenchStore.getState();
    expect(updatedState.widgets[constructKey].id).toEqual(expected);
  });
  it("doesn't foobar anything else", () => {
    const constructKey = 'whatever';
    useWorkbenchStore.setState({
      widgets: {
        [constructKey]: { id: '', x: 0, y: 0, forType: 'whatever' },
        asdfasdf: { id: 'asdfasdf', x: 1, y: 1, forType: 'somethingelse' },
      },
      pickers: {},
    });

    // when
    const expected = 'something';
    useWorkbenchStore.getState().updateConstructId(constructKey, expected);

    // then
    const updatedState: WorkbenchState = useWorkbenchStore.getState();
    expect(updatedState.widgets).toMatchSnapshot();
  });
});

describe('addWidget', () => {
  it('adds and calls update', () => {
    const constructKey = 'whatever';
    useWorkbenchStore.setState({
      widgets: {
        [constructKey]: {
          id: '',
          initializerParameters: {},
          x: 0,
          y: 0,
          forType: 'somefqnname',
        },
      },
      pickers: {},
    });
    let newKeyCreated = '';
    const newWidget = {};

    useWorkbenchStore
      .getState()
      .addWidget(
        newWidget as WidgetState,
        (state: WorkbenchState, newKey: string) => {
          newKeyCreated = newKey;
          if (
            !(state.widgets[constructKey] as ConstructWidgetState)
              .initializerParameters
          ) {
            (
              state.widgets[constructKey] as ConstructWidgetState
            ).initializerParameters = {};
          }
          (
            state.widgets[constructKey] as ConstructWidgetState
          ).initializerParameters!.something = {
            ref: getRef(newKey),
            widgetShapes: {},
          };
        }
      );

    const updatedScope = useWorkbenchStore.getState();
    expect(
      (updatedScope.widgets[constructKey] as ConstructWidgetState)
        .initializerParameters.something.ref
    ).toEqual(getRef(newKeyCreated));
    expect(updatedScope.widgets[newKeyCreated]).toEqual(newWidget);
  });

  it('work without updater', () => {
    const constructKey = 'whatever';
    useWorkbenchStore.setState({
      widgets: {
        [constructKey]: {
          id: '',
          initializerParameters: {},
          x: 0,
          y: 0,
          forType: 'somefqnname',
        },
      },
      pickers: {},
    });

    useWorkbenchStore.getState().addWidget({} as WidgetState);

    const updatedScope = useWorkbenchStore.getState();
    expect(Object.values(updatedScope.widgets)).toHaveLength(2);
  });
});

describe('updateWidgets', function () {
  it('should set values', function () {
    const staticFunctionKey = 'whatever';
    useWorkbenchStore.setState({
      widgets: {
        [staticFunctionKey]: {
          id: '',
          parameters: {},
          x: 0,
          y: 0,
          forType: 'somefqnname',
        },
      },
      pickers: {},
    });

    useWorkbenchStore.getState().updateWidget((widgets) => {
      (
        widgets[staticFunctionKey] as StaticFunctionWidgetState
      ).parameters.whatever = { value: 'thing' };
    });
    const updatedState = useWorkbenchStore.getState();
    expect(
      (updatedState.widgets[staticFunctionKey] as StaticFunctionWidgetState)
        .parameters.whatever
    ).toEqual({ value: 'thing' });
  });
});
