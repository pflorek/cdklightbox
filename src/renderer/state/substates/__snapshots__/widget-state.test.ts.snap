// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`update construct id doesn't foobar anything else 1`] = `
Object {
  "asdfasdf": Object {
    "forType": "somethingelse",
    "id": "asdfasdf",
    "x": 1,
    "y": 1,
  },
  "whatever": Object {
    "forType": "whatever",
    "id": "something",
    "x": 0,
    "y": 0,
  },
}
`;
